modules/layers/src/text-layer/font-atlas-manager.js

```diff
@@ -0,0 +1,264 @@
+    }
+
+    // update fontAtlas with new settings
+    const fontAtlas = this._generateFontAtlas(this._key, charSet, cachedFontAtlas);
+    this._updateTexture(fontAtlas);
+
+    // update cache
+    cache.set(this._key, fontAtlas);
+  }
+
+  _updateTexture({data: canvas, width, height}) {
+    // resize texture
+    if (this._texture.width !== width || this._texture.height !== height) {
+      this._texture.resize({width, height});
+    }
+
+    // update image data
+    this._texture.setImageData({
+      data: canvas,
+      width,
+      height,
+      parameters: {
+        [GL_TEXTURE_WRAP_S]: GL_CLAMP_TO_EDGE,
+        [GL_TEXTURE_WRAP_T]: GL_CLAMP_TO_EDGE,
+        [GL.UNPACK_FLIP_Y_WEBGL]: true
+      }
+    });
+
+    // this is required step after texture data changed
+    this._texture.generateMipmap();
+  }
+
+  _generateFontAtlas(key, characterSet, cachedFontAtlas) {
+    const {fontFamily, fontWeight, fontSize, buffer, sdf, radius, cutoff} = this;
+    let canvas = cachedFontAtlas && cachedFontAtlas.data;
+    if (!canvas) {
+      canvas = document.createElement('canvas');
+      canvas.width = MAX_CANVAS_WIDTH;
+    }
+    const ctx = canvas.getContext('2d');
+
+    setTextStyle(ctx, fontFamily, fontSize, fontWeight);
+
+    // 1. build mapping
+    const {mapping, canvasHeight, xOffset, yOffset} = buildMapping(
+      Object.assign(
+        {
+          getFontWidth: char => ctx.measureText(char).width,
+          fontHeight: fontSize * HEIGHT_SCALE,
+          buffer,
+          characterSet,
+          maxCanvasWidth: MAX_CANVAS_WIDTH
+        },
+        cachedFontAtlas && {
+          mapping: cachedFontAtlas.mapping,
+          xOffset: cachedFontAtlas.xOffset,
+          yOffset: cachedFontAtlas.yOffset
+        }
+      )
+    );
+
+    // 2. update canvas
+    // copy old canvas data to new canvas only when height changed
+    if (canvas.height !== canvasHeight) {
+      const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
+      canvas.height = canvasHeight;
+      ctx.putImageData(imageData, 0, 0);
+    }
+    setTextStyle(ctx, fontFamily, fontSize, fontWeight);
+
+    // 3. layout characters
+    if (sdf) {
+      const tinySDF = new TinySDF(fontSize, buffer, radius, cutoff, fontFamily, fontWeight);
+      // used to store distance values from tinySDF
+      // tinySDF.size equals `fontSize + buffer * 2`
+      const imageData = ctx.getImageData(0, 0, tinySDF.size, tinySDF.size);
+
+      for (const char of characterSet) {
+        populateAlphaChannel(tinySDF.draw(char), imageData);
+        ctx.putImageData(imageData, mapping[char].x - buffer, mapping[char].y - buffer);
+      }
+    } else {
+      for (const char of characterSet) {
+        ctx.fillText(char, mapping[char].x, mapping[char].y + fontSize * BASELINE_SCALE);
+      }
+    }
+
+    return {
+      xOffset,
+      yOffset,
+      mapping,
+      data: canvas,
+      width: canvas.width,
+      height: canvas.height
+    };
+  }
```

Comment:
return object structure changed.

old code 

```
  return {
    scale: HEIGHT_SCALE,
    mapping,
    texture: new Texture2D(gl, {
      pixels: canvas,
      // padding is added only between the characters but not for borders
      // enforce CLAMP_TO_EDGE to avoid any artifacts.
      parameters: {
        [GL_TEXTURE_WRAP_S]: GL_CLAMP_TO_EDGE,
        [GL_TEXTURE_WRAP_T]: GL_CLAMP_TO_EDGE
      }
    })
  };
```